core/tests/unit/kmnkbd/test_actions_normalize.cpp

```diff
@@ -118,44 +131,115 @@ void test(
+ *                                      actions_output) - markers supported
+ * @param actions_code_points_to_delete number of NFU code points that the
+ *                                      keyboard processor has asked to remove
+ *                                      in its actions
+ * @param actions_output                NFU string that the keyboard processor
+ *                                      has asked to insert in its actions
+ * @param expected_delete               expected: NFU code points to ask app to
+ *                                      remove
+ * @param expected_output               expected: adjusted NFU output to insert
+ *                                      into the app
+ * @param expected_final_app_context    expected: NFU adjusted final app
+ *                                      context, which will be NFC from the
+ *                                      boundary of the transform, but will not
+ *                                      have been modified prior to that. Should
+ *                                      match char-for-char what the app ends up
+ *                                      with in its text buffer.
+ */
+void test_actions_update_app_context_nfu(
+  const char *name,
+  const km_core_cp *initial_app_context,
+  const km_core_cp *final_cached_context_string,
+  const km_core_context_item *final_cached_context_items,
+  int actions_code_points_to_delete,
+  const std::u32string actions_output,
+
+  const unsigned int expected_delete,
+  const std::u32string expected_output,
+  const km_core_cp *expected_final_app_context
+) {
+  std::cout << "test_actions_update_app_context_nfu: " << name << std::endl;
+
+  setup(initial_app_context, final_cached_context_string, final_cached_context_items, actions_code_points_to_delete, actions_output);
+
+  assert(km::core::actions_update_app_context_nfu(km_core_state_context(test_state), km_core_state_app_context(test_state)));
+
+  std::cout << "test_actions_update_app_context_nfu: (" << name << "): delete: " << expected_delete << " output: |" << std::u32string(test_actions.output) << "|" << std::endl;
+  std::u32string o(test_actions.output);
+  for(auto i = o.begin(); i < o.end(); i++) {
+    std::cout << "U+" << std::hex << (int)(*i) << " ";
+  }
+  std::cout << std::endl;
+
+  assert(expected_delete == test_actions.code_points_to_delete);
+  assert(expected_output == test_actions.output);
+
+  auto debug = km_core_state_context_debug(test_state, KM_CORE_DEBUG_CONTEXT_APP);
+  std::cout << " final app context: " << debug << std::endl;
+  km_core_cp_dispose(debug);
+
+  compare_context(km_core_state_app_context(test_state), expected_final_app_context);
+  teardown();
+}
+
+void run_actions_normalize_tests() {
 
   // Null boundary tests
 
-  test(
+  test_actions_normalize(
     "noop",
     /* app context pre transform: */     u"",
     /* cached context post transform: */ u"",
+    /* cached context post transform: */ nullptr,
```

Comment:
It took me a while to realise that this is a one or the other situation. The caller either provides a cached context string XOR a cached context items array. At least I think that is the case. Maybe have a comment about this where `test_actions_normalize` is defined.

Reply:
https://github.com/keymanapp/keyman/pull/10607/files#r1477694532